enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+
+TODO: How to not fall into endless loop of rollbacks 
+
+### Integration with greenboot
+
+Depending on result of greenboot's healthcheck either "green" (successful boot) or "red" (unsuccessful) scripts are executed before rebooting the system.
+
+MicroShift will integrate with that system to persist an action to perform on next boot:
+- "green": on next boot, before MicroShift starts, make a backup of MicroShift's data
+- "red": on next boot, before MicroShift starts, restore MicroShift's data from backup
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
+
+File containing said information should not be part of MicroShift data directory
+as we don't to be a part of backup.
+
+### Backup and restore of MicroShift data
+
+Decision whether to backup or restore is based on file persisted during previous boot (see "Integration with greenboot").
+When file is read to make a decision (but not during dry run), it shall be removed.
+
+As a result of investigation and aiming for simplicity for initial implementation,
+it was decided that backing up MicroShift's data will be done by leveraging using copy-on-write (CoW) functionality.
+
+CoW is a feature of filesystem (supported by XFS and Btrfs) and it can be used by providing a `--reflink=` param to `cp` option.
+`--reflink=auto` will be used over `--reflink=always` to gracefully fall back to regular copying on filesystems
+not supporting CoW (ext4, ZFS).
+Backup will be done in `/var/lib/microshift.bak` or similar - it needs to be within the same filesystem/mount.
+Only one backup will be kept. Consequent boots will overwrite the backup dir (only if the action is "backup").
+
+End user documentation needs to include guidance on setting up filesystem to fullfil requirements for using copy-on-write (e.g. making sure some filesystem options are not disabled).
+
+### MicroShift version metadata persistence
+
+When MicroShift is up and running healthy, it will persist its own version and ID of current ostree commit into data dir as a file, e.g.:
+```json
+{
+  "microshift": "4.14.0",
+  "ostree_commit": "1a2b",
+}
+```
+
+MicroShift version will be used on consequent reboots to determine:
+- is upgrade is allowed or blocked
+- should storage migration be performed
+
+ID of ostree commit will be used to... TODO: Think if we can use it for "System rollback and failed data restore" scenario
+
+### Allowing and blocking upgrades (change of MicroShift version)
+
+Decision to allow or block an upgrade (change) will be based on persisted MicroShift version.
```

Comment:
This list might be clearer if the checks focused on when we would _not_ allow the upgrade, so that if you make it to the end of the list then the action is to proceed with the update. The order probably matters, too, right? We need to determine the current version (including applying the assumption from line 421), check the skew, check the explicit block list, decide whether to run storage migration, then start the full server.

Reply:
Please, check new version